File: Python/204.count-primes.py
#
# @lc app=leetcode id=204 lang=python
#
# [204] Count Primes
#
# https://leetcode.com/problems/count-primes/description/
#
# algorithms
# Easy (30.07%)
# Likes:    1380
# Dislikes: 464
# Total Accepted:    284.1K
# Total Submissions: 942.4K
# Testcase Example:  '10'
#
# Count the number of prime numbers less than a non-negative number, n.
# 
# Example:
# 
# 
# Input: 10
# Output: 4
# Explanation: There are 4 prime numbers less than 10, they are 2, 3, 5, 7.
# 
# 
#

# @lc code=start
class Solution(object):
    def countPrimes(self, n):
        """
        :type n: int
        :rtype: int
        """

        if n <= 2:
            return 0
        res = [True] * n
        res[0] = res[1] = False
        for i in range(2, n):
            if res[i]:
                for j in range(2, (n-1)//i+1):
                    res[i*j] = False    
        return sum(res)
        
# @lc code=end

File: C++/204.count-primes.cpp
/*
 * @lc app=leetcode id=204 lang=cpp
 *
 * [204] Count Primes
 *
 * https://leetcode.com/problems/count-primes/description/
 *
 * algorithms
 * Easy (30.47%)
 * Likes:    1555
 * Dislikes: 511
 * Total Accepted:    304.4K
 * Total Submissions: 998.3K
 * Testcase Example:  '10'
 *
 * Count the number of prime numbers less than a non-negative number, n.
 * 
 * Example:
 * 
 * 
 * Input: 10
 * Output: 4
 * Explanation: There are 4 prime numbers less than 10, they are 2, 3, 5, 7.
 * 
 * 
 */
#include<vector>
#include<numeric>
#include<math.h>
using namespace std;

// @lc code=start
class Solution {
public:
    int countPrimes(int n) {
         if (n<2) return 0;
        
        vector<int> primes(n, 1);
        primes[0] = primes[1] = 0;

        for (int i=2; i<sqrt(n); i++) {
            if (primes[i]) {
                for(int j=2; i*j<n; j++) 
                    primes[i*j] = 0;
            }
        }
        return accumulate(primes.begin(), primes.end(), 0);
        return count(primes.begin(), primes.end(), 1);    
 
    }
};
// @lc code=end

